### PR TITLE
overhaul Twilight Princess

### PR DESCRIPTION
**Pull request by TakaRikka**
This overhauls the TP setup to be more consistent with the TWW setup, while also adding a bunch of stuff like actors and the environment system. There's still quite a lot to fix and add before it's in a state I'm happy with, but I think it's a pretty major improvement over what was there before and I need to do more decomp work to research more things.

Notes:
- The environment system still is relatively broken, but it's at least in a presentable state. Actor / BG lighting needs a lot of work that I need to do more decomp work on. Most weather packets aren't working yet.
- For simplicities sake I made the extractor use TP decomp's symbol map files from a clean matching build. The original symbol maps lack a lot of information that makes parsing it a lot more annoying. If this is a problem it can still probably be fixed, but this was much less effort

### CODE REVIEW

File: src/TwilightPrincess/d_kankyo.ts [hunk 1]
```diff
@@ -0,0 +1,2150 @@
...
+    [5, 10, 5],
+    [7, 15, 5],
+    [5, 12.5, 5],
+    [10],
+];
+
+function dice_wether_init(envLight: dScnKy_env_light_c, mode: DiceWeatherMode, timeChange: number, time: number): void {
+    console.log(`d_kankyo: dice_wether_init`, DiceWeatherMode[mode]);
+
+    envLight.diceWeatherMode = mode;
```
Comment (magcius) [inline src/TwilightPrincess/d_kankyo.ts:1383]: I'm curious, is the dice weather stuff the same, with the same patterns? Or was this copy/paste

File: src/TwilightPrincess/d_kankyo.ts [hunk 2]
```diff
@@ -0,0 +1,2150 @@
...
+
+    envLight.timeSpeed = 0.012;
+
+    colorFromRGBA(envLight.lightStatus[0].Color, 1.0, 0.0, 0.0, 0.0);
+    colorFromRGBA(envLight.lightStatus[1].Color, 0.0, 0.0, 0.0, 0.0);
+
+    envLight.diceWeatherChangeTime = (envLight.curTime + 15.0) % 360.0;
+
+    if ((today.getDay() === 5 && today.getDate() === 13) || (today.getMonth() === 9 && today.getDate() === 31))
+        envLight.eventNightStop = true;
```
Comment (magcius) [inline src/TwilightPrincess/d_kankyo.ts:1823]: this is an easter egg in Wind Waker :)

I don't know if eventNightStop has any meaningful impact in TP

File: src/TwilightPrincess/d_kankyo.ts [hunk 3]
```diff
@@ -0,0 +1,2150 @@
...
+    color.g = g * ratio * 1/255;
+    color.b = b * ratio * 1/255;
+}
+
+export function dKy_fog_startendz_set(envLight: dScnKy_env_light_c, param_0: number, param_1: number, ratio: number): void {
+    if (ratio < 0 || ratio > 1) {
+        ratio = 0;
+    }
+
+    if (ratio < 9.999999747378752e-05) {
```
Comment (magcius) [inline src/TwilightPrincess/d_kankyo.ts:1837]: 0.0001

File: src/TwilightPrincess/d_kankyo.ts [hunk 4]
```diff
@@ -0,0 +1,2150 @@
...
+        envLight.BGpartsActiveLight[index] = envLight.eflights[index];
+        envLight.BGpartsActiveLight[index].index = index + 1;
+    }
+}
+
+export function dKy_bgparts_activelight_cut(envLight: dScnKy_env_light_c, index: number): void {
+    envLight.BGpartsActiveLight[index].index = 0;
+}
+
+const lightMaskData: number[] = [1, 2, 4, 8, 16, 32, 64, 128];
```
Comment (magcius) [inline src/TwilightPrincess/d_kankyo.ts:1972]: In the original code, this is probably GX_LIGHT0, GX_LIGHT1, etc.

We can rely on it just being the bitfields, so I would just replace `lightMaskData[n]` with `(1 << n)`

File: src/TwilightPrincess/d_kankyo_wether.ts [hunk 5]
```diff
@@ -0,0 +1,2258 @@
...
+/*
+                const ctx = getDebugOverlayCanvas2D();
+                const c = textureIdx === 0 ? Magenta : textureIdx === 1 ? Green : Blue;
+                colorCopy(materialParams.u_Color[ColorKind.C0], c, kumo.alpha);
+                vec3.set(scratchVec3e, x, y, z);
+                vec3.scale(scratchVec3e, scratchVec3e, 10000);
+                scratchVec3e[0] += -196400;
+                scratchVec3e[1] = scratchVec3e[1] * 0.5 + 3000;
+                scratchVec3e[2] += 295960;
+                drawWorldSpacePoint(ctx, viewerInput.camera.clipFromWorldMatrix, scratchVec3e, materialParams.u_Color[ColorKind.C0], 50 * height);
```
Comment (magcius) [inline src/TwilightPrincess/d_kankyo_wether.ts:925]: hahaha you can get rid of this, sorry, this was tweet draw code

File: src/TwilightPrincess/d_kankyo_wether.ts [hunk 6]
```diff
@@ -0,0 +1,2258 @@
...
+        vec3.fromValues(3906, 31005, -23144),
+        vec3.fromValues(-439, 30127, -17919),
+        vec3.fromValues(-7421, 31005, 18798),
+        vec3.fromValues(-10937, 2800, 15000),
+        vec3.fromValues(-10000, 24902, 18400),
+        vec3.fromValues(-9400, 22500, 15900),
+        vec3.fromValues(-9179, 21300, 14300),
+        vec3.fromValues(-10300, 22000, 21000),
+        vec3.fromValues(-16000, 25500, 20000),
+        vec3.fromValues(0, 30000, 19000),
```
Comment (magcius) [inline src/TwilightPrincess/d_kankyo_wether.ts:1309]: This is the big dipper, which I can't imagine they kept for Twilight Princess.

File: src/TwilightPrincess/d_resorce.ts [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        if (entry === null)
+            return null;
+        return this.lazyLoadResource(resType, entry);
+    }
+
+    private autoLoadResource(device: GfxDevice, cache: GfxRenderCache, type: string, resEntry: ResEntry<any>): void {
+        const file = resEntry.file;
+        if (type === `BMD ` || type === `BMDM` || type === `BMDC` || type === `BMDS` || type === `BSMD` ||
+            type === `BDL ` || type === `BDLM` || type === `BDLC` || type === `BDLI` || type === `BMDE` ||
+            type === `BMDP` || type === `BMDR` || type === `BMDV` || type === `BMWR` || type === `BMDG`) {
```
Comment (magcius) [inline src/TwilightPrincess/d_resorce.ts:206]: zoop

File: src/TwilightPrincess/m_do_ext.ts [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    public update(): void {
+        this.model.jointMatrixCalc = this;
+    }
+
+    public entryDL(globals: dGlobals, renderInstManager: GfxRenderInstManager, viewerInput: ViewerRenderInput, drawListSet: dDlst_list_Set | null = null): void {
+        mDoExt_modelEntryDL(globals, this.model, renderInstManager, viewerInput);
+    }
+}
```
Comment (magcius) [inline src/TwilightPrincess/m_do_ext.ts:226]: I think we can share this right now, but I'm certain that won't be the case in the future, so I don't mind leaving this one separate.

File: src/TwilightPrincess/d_a.ts [hunk 9]
```diff
@@ -0,0 +1,1740 @@
...
+
+        for (let i = 0; i < this.numBg; i++) {
+            if (this.bgModel[i] === null)
+                continue;
+
+            settingTevStruct(globals, LightType.BG0 + i, null, this.bgTevStr[i]!);
+            setLightTevColorType(globals, this.bgModel[i]!, this.bgTevStr[i]!, viewerInput.camera);
+            // this is actually mDoExt_modelEntryDL
+
+            const m2 = this.bgModel[i]!.getTextureMappingReference('fbtex_dummy');
```
Comment (magcius) [inline src/TwilightPrincess/d_a.ts:262]: interesting, can it really put all 6 models into indirect if it wants?

File: src/TwilightPrincess/ztp_scenes.ts [hunk 10]
```diff
@@ -0,0 +1,1182 @@
...
+
+        // Grass/Flowers/Trees
+        const frameCount = viewerInput.time / 1000.0 * 30;
+
+        fopDw_Draw(globals.frameworkGlobals, globals, renderInstManager, viewerInput);
+
+        // TODO: fix bloom, errors atm
+        /* globals.bloom_c.enable = true;
+        globals.bloom_c.monoColor = colorNewCopy(White);
+        globals.bloom_c.draw(globals, renderInstManager, viewerInput); */
```
Comment (magcius) [inline src/TwilightPrincess/ztp_scenes.ts:864]: I can help with this if needed

**Review by magcius — APPROVED** [review]
Lots of notes here. As always, let me know if you have any questions. Feel free to ignore any suggestion here, I'm OK with merging as-is.

File: src/TwilightPrincess/SComponent.ts [hunk 11]
```diff
@@ -0,0 +1,161 @@
+
+import { ReadonlyVec3, vec3 } from "gl-matrix";
+import { clamp, MathConstants, normToLength } from "../MathHelpers.js";
+
+function clampAbs(v: number, min: number, max: number): number {
```
[the remark this answers is not in the document]
Comment (TakaRikka) [reply src/TwilightPrincess/SComponent.ts:5]: for all these wind waker copied files, would you prefer to move them to a more generic folder or just have tp include the files from the wind waker folder?

File: src/TwilightPrincess/ztp_scenes.ts (hunk 10 above)
Comment (TakaRikka) [reply src/TwilightPrincess/ztp_scenes.ts:864]: yeah i'm still pretty lost on figuring out the GX calls atm

File: src/TwilightPrincess/d_kankyo_wether.ts (hunk 6 above)
Comment (TakaRikka) [reply src/TwilightPrincess/d_kankyo_wether.ts:1309]: surprisingly this actually is in tp, still used in dKyr_drawStar

File: src/TwilightPrincess/SComponent.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (magcius) [reply src/TwilightPrincess/SComponent.ts]: just have tp include from ww for now

File: src/TwilightPrincess/d_kankyo_wether.ts (hunk 6 above)
Comment (magcius) [reply src/TwilightPrincess/d_kankyo_wether.ts:1297]: lmao, ok then!

**Pull request merged**